cocos/physics/cannon/cannon-rigid-body.ts

```diff
@@ -31,23 +31,20 @@ export class CannonRigidBody implements IRigidBody {
     }
 
     setAllowSleep (v: boolean) {
-        if (this.impl.isSleeping()) {
-            this.impl.wakeUp();
-        }
+        this.impl.wakeUp();
         this.impl.allowSleep = v;
     }
 
     setMass (value: number) {
         this.impl.mass = value;
         if (this.impl.mass == 0) {
             this.impl.type = CANNON.Body.STATIC;
+        } else {
+            this.impl.type = this._rigidBody.isKinematic ? CANNON.Body.KINEMATIC : CANNON.Body.DYNAMIC;
         }
 
         this.impl.updateMassProperties();
-
-        if (this.impl.isSleeping()) {
-            this.impl.wakeUp();
-        }
+        this.impl.wakeUp();
```

Comment:
Is the lower level implementation checking state so that it won't run logic repetitively?

Reply:
Great advice, it should check indeed.